### PR TITLE
Endpoint for eventgridevents (deliveryreport)

### PR DESCRIPTION
**Pull request by tba76**
## Related Issue(s)
- Altinn/altinn-notifications#281

## Verification
- [ ] **Your** code builds clean without any errors or warnings
- [ ] Manual testing done (required)
- [ ] Relevant automated test added (if you find this hard, leave it and we'll help out)
- [ ] All tests run green

## Documentation
- [ ] User documentation is updated with a separate linked PR in [altinn-studio-docs.](https://github.com/Altinn/altinn-studio-docs) (if applicable)

### CODE REVIEW

File: src/Altinn.Notifications.Email.Core/Status/EmailSendResultMapper.cs [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+{
+    /// <summary>
+    /// Parse AcsEmailDeliveryReportStatus to EmailSendResult
+    /// </summary>
+    /// <param name="deliveryStatus">Delivery status from Azure Communication Service</param>
+    /// <returns></returns>
+    /// <exception cref="ArgumentException">Throws exception if unknown delivery status</exception>
+    public static EmailSendResult ParseDeliveryStatus(AcsEmailDeliveryReportStatus? deliveryStatus)
+    {
+        if (deliveryStatus == null)
```
Comment (acn-sbuad) [inline src/Altinn.Notifications.Email.Core/Status/EmailSendResultMapper.cs:18]: Would a switch be easier to read here?

Comment (acn-sbuad) [reply src/Altinn.Notifications.Email.Core/Status/EmailSendResultMapper.cs:18]: Alternatively a dictionary and a lookup would be less lines of code.

File: src/Altinn.Notifications.Email.Core/Altinn.Notifications.Email.Core.csproj [hunk 2]
```diff
@@ -9,6 +9,7 @@
   </PropertyGroup>
 
   <ItemGroup>
+    <PackageReference Include="Azure.Messaging.EventGrid" Version="4.22.0" />
```
Comment (acn-sbuad) [inline src/Altinn.Notifications.Email.Core/Altinn.Notifications.Email.Core.csproj:12]: not a fan of this event grid dependency on core, but I guess there is not better way to implement this?

Comment (tba76) [reply src/Altinn.Notifications.Email.Core/Altinn.Notifications.Email.Core.csproj:12]: I agree, but we are just using the models to simplify event-handling from eventgrid.

File: src/Altinn.Notifications.Email.Core/Status/EmailSendResultMapper.cs (hunk 1 above)
Comment (tba76) [reply src/Altinn.Notifications.Email.Core/Status/EmailSendResultMapper.cs:18]: Changed to switch with a little workaround

File: src/Altinn.Notifications.Email.Core/Status/EmailSendResultMapper.cs [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+/// </summary>
+public static class EmailSendResultMapper
+{
+    /// <summary>
+    /// Parse AcsEmailDeliveryReportStatus to EmailSendResult
+    /// </summary>
+    /// <param name="deliveryStatus">Delivery status from Azure Communication Service</param>
+    /// <returns></returns>
+    /// <exception cref="ArgumentException">Throws exception if unknown delivery status</exception>
+    public static EmailSendResult ParseDeliveryStatus(string? deliveryStatus)
```
Comment (acn-sbuad) [inline src/Altinn.Notifications.Email.Core/Status/EmailSendResultMapper.cs:14]: I guess this mapper could be moved to the API project as it is the only place the mapper is used.

File: src/Altinn.Notifications.Email/Configuration/AuthSettings.cs [hunk 4]
```diff
@@ -0,0 +1,12 @@
+﻿namespace Altinn.Notifications.Email.Configuration;
+
+/// <summary>
+/// Configuration object used to hold authorization settings for the delivery report controller.
+/// </summary>
+public class AuthSettings
```
Comment (acn-sbuad) [inline src/Altinn.Notifications.Email/Configuration/AuthSettings.cs:6]: Did we want to go for the full name here? 

```suggestion
public class AuthorizationSettings
```

File: src/Altinn.Notifications.Email/Attributes/AccessKeyAttribute.cs [hunk 5]
```diff
@@ -0,0 +1,29 @@
...
+/// </summary>
+[AttributeUsage(AttributeTargets.Class | AttributeTargets.Method)]
+public class AccessKeyAttribute : Attribute, IAsyncActionFilter
+{
+    /// <summary>
+    /// Method that is called before the action is executed.
+    /// </summary>
+    public async Task OnActionExecutionAsync(ActionExecutingContext context, ActionExecutionDelegate next)
+    {
+        AuthSettings authSettings = context.HttpContext.RequestServices.GetRequiredService<IConfiguration>()!.GetSection("AuthSettings").Get<AuthSettings>()!;
```
Comment (acn-sbuad) [inline src/Altinn.Notifications.Email/Attributes/AccessKeyAttribute.cs:18]: We may need a longer section name to match in the environment vars. AuthSettings__Accesskey is a secret name that might drown a bit in key vault amongst all other secrets?

File: src/Altinn.Notifications.Email.Core/Status/EmailSendResultMapper.cs [hunk 6]
```diff
@@ -0,0 +1,36 @@
...
+    {
+        switch (deliveryStatus)
+        {
```
Comment (acn-sbuad) [inline src/Altinn.Notifications.Email.Core/Status/EmailSendResultMapper.cs:17]: Wondering how we persist the information about an unknown status being pushed to us. Should we log the operationId and status so we can follow up and potentially manually update the database? Currently I guess only the status is logged, but we won't know for which sms

File: src/Altinn.Notifications.Email/Controllers/DeliveryReportController.cs [hunk 7]
```diff
@@ -0,0 +1,69 @@
...
+    public async Task<ActionResult<string>> Post([FromBody] EventGridEvent[] eventList)
+    {
+        foreach (EventGridEvent eventgridevent in eventList)
+        {
+            // If the event is a system event, TryGetSystemEventData will return the deserialized system event
+            if (eventgridevent.TryGetSystemEventData(out object systemEvent))
+            {
+                switch (systemEvent)
+                {
+                    case SubscriptionValidationEventData subscriptionValidated:
```
Comment (acn-sbuad) [inline src/Altinn.Notifications.Email/Controllers/DeliveryReportController.cs:49]: would this be a suitable place to comment WHY we return the responsdata ? I guess people can se what is done, but it is not intuitive why?

File: src/Altinn.Notifications.Email/Configuration/AuthSettings.cs (hunk 4 above)
Comment (tba76) [reply src/Altinn.Notifications.Email/Configuration/AuthSettings.cs:6]: Changed to EmailDeliveryReportSettings

File: src/Altinn.Notifications.Email/Controllers/DeliveryReportController.cs (hunk 7 above)
Comment (tba76) [reply src/Altinn.Notifications.Email/Controllers/DeliveryReportController.cs:49]: Done

File: src/Altinn.Notifications.Email.Core/Status/EmailSendResultMapper.cs (hunk 3 above)
Comment (tba76) [reply src/Altinn.Notifications.Email.Core/Status/EmailSendResultMapper.cs:14]: Moved

File: src/Altinn.Notifications.Email.Core/Status/EmailSendResultMapper.cs (hunk 6 above)
Comment (tba76) [reply src/Altinn.Notifications.Email.Core/Status/EmailSendResultMapper.cs]: Added logging in the controller

**Review by acn-sbuad — APPROVED** [review]
🎈

**Pull request merged**